frontend/catalyst/qfunc.py

```diff
@@ -129,10 +128,8 @@ def __call__(self, *args, **kwargs):
         device_capabilities = get_device_capabilities(self.device, program_features)
         backend_info = QFunc.extract_backend_info(self.device, device_capabilities)
 
-        if isinstance(self.device, qml.devices.Device):
-            qjit_device = QJITDeviceNewAPI(self.device, device_capabilities, backend_info)
-        else:
-            qjit_device = QJITDevice(self.device, device_capabilities, backend_info)
+        assert isinstance(self.device, qml.devices.Device), "Unsupported device API"
```

Comment:
This could feasibly be triggered by a user so an exception would be better

Reply:
Done

Reply:
It seems the proposed exception will hardly be reached because Pennylane has removed the legacy `qml.Device` interface and replaced it with an abstract class called `qml.devices.LegacyDevice`. On top of that, the new `qml.device` converts all Legacy Devices to the new API with the `LegacyDeviceFacade` class. Moreover, a Custom Legacy Device derived from the abstract class would require also a toml specification and a shared object in order to work with Catalyst. We are deleting the documentation section that describes how to use the legacy API.

In short, to test the proposed exception we would require a legacy device with toml spec at least, so we pass the compilation constrains. Is it worth doing that for a case that is very unlikely to happen? 

I would say: let's turn into an assertion in the meantime.

Reply:
Can we just remove it?

Reply:
From what I understand, we still support the old device API through the facade.

Reply:
I agree with Raul.

Reply:
But the `LegacyDeviceFacade `type is in fact an instance of `qml.devices.Device`, hence the condition for the exception will never be met:
```
if not isinstance(self.device, qml.devices.Device):
  raise TypeError("Unsupported device API. Please use 'qml.devices.Device' instead.")
```